cpp/src/arrow/filesystem/test-util.cc

```diff
@@ -0,0 +1,704 @@
+  ASSERT_OK(fs->CreateDir("EF"));
+  CreateFile(fs, "AB/abc", "data");
+  std::vector<std::string> all_dirs{"AB", "AB/CD", "EF"};
+
+  // Copy into root dir
+  ASSERT_OK(fs->CopyFile("AB/abc", "def"));
+  AssertAllDirs(fs, all_dirs);
+  AssertAllFiles(fs, {"AB/abc", "def"});
+
+  // Copy out of root dir
+  ASSERT_OK(fs->CopyFile("def", "EF/ghi"));
+  AssertAllDirs(fs, all_dirs);
+  AssertAllFiles(fs, {"AB/abc", "EF/ghi", "def"});
+
+  // Overwrite contents for one file => other data shouldn't change
+  CreateFile(fs, "def", "other data");
+  AssertFileContents(fs, "AB/abc", "data");
+  AssertFileContents(fs, "def", "other data");
+  AssertFileContents(fs, "EF/ghi", "data");
+
+  // Destination is a file => clobber
+  ASSERT_OK(fs->CopyFile("def", "AB/abc"));
+  AssertAllDirs(fs, all_dirs);
+  AssertAllFiles(fs, {"AB/abc", "EF/ghi", "def"});
+  AssertFileContents(fs, "AB/abc", "other data");
+  AssertFileContents(fs, "def", "other data");
+  AssertFileContents(fs, "EF/ghi", "data");
+
+  // Identical source and destination: allowed to succeed or raise IOError,
+  // but should not lose data.
+  Status st = fs->CopyFile("def", "def");
+  if (!st.ok()) {
+    ASSERT_RAISES(IOError, st);
+  }
+  AssertAllFiles(fs, {"AB/abc", "EF/ghi", "def"});
+  AssertFileContents(fs, "def", "other data");
+
+  // Destination is a directory
+  ASSERT_RAISES(IOError, fs->CopyFile("def", "AB"));
+  // Source doesn't exist
+  ASSERT_RAISES(IOError, fs->CopyFile("abc", "xxx"));
+  // Parent destination doesn't exist
+  ASSERT_RAISES(IOError, fs->Move("AB/abc", "XX/mno"));
+  // Parent destination is not a directory
+  ASSERT_RAISES(IOError, fs->Move("AB/abc", "def/mno"));
+  AssertAllDirs(fs, all_dirs);
+  AssertAllFiles(fs, {"AB/abc", "EF/ghi", "def"});
+}
+
+void GenericFileSystemTest::TestGetTargetStatsSingle(FileSystem* fs) {
+  ASSERT_OK(fs->CreateDir("AB/CD"));
+  CreateFile(fs, "AB/CD/ghi", "some data");
+
+  FileStats st;
+  TimePoint first_time;
+  ASSERT_OK(fs->GetTargetStats("AB", &st));
+  AssertFileStats(st, "AB", FileType::Directory);
+  ASSERT_EQ(st.base_name(), "AB");
+  ASSERT_EQ(st.size(), kNoSize);
+  first_time = st.mtime();
+  ValidateTimePoint(first_time);
+
+  ASSERT_OK(fs->GetTargetStats("AB/CD", &st));
+  AssertFileStats(st, "AB/CD", FileType::Directory);
+  ASSERT_EQ(st.base_name(), "CD");
+  ASSERT_EQ(st.size(), kNoSize);
+  // AB/CD was created a little after AB
+  AssertDurationBetween(st.mtime() - first_time, 0.0, kTimeSlack);
+
+  ASSERT_OK(fs->GetTargetStats("AB/CD/ghi", &st));
+  AssertFileStats(st, "AB/CD/ghi", FileType::File, 9);
+  ASSERT_EQ(st.base_name(), "ghi");
+  AssertDurationBetween(st.mtime() - first_time, 0.0, kTimeSlack);
+
+  ASSERT_OK(fs->GetTargetStats("zz", &st));
+  AssertFileStats(st, "zz", FileType::NonExistent);
+  ASSERT_EQ(st.base_name(), "zz");
+  ASSERT_EQ(st.size(), kNoSize);
+  ASSERT_EQ(st.mtime(), kNoTime);
+
+  // Invalid path
+  // XXX will this really be rejected by all filesystems?
```

Comment:
Debatable. Note we also have to care about Windows filesystem. Speaking of Windows, are you thinking about translating user paths into Windows paths? It would be nice to work with just forward slashes :/

I've seen some weird file paths for accessing network directories inside large enterprises so this may be a place we end up spending some time

Reply:
Right, we may need a translation layer between Windows paths and URI-like paths.